{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "import data_utils\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/tcga_survival_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = pd.read_csv('data/cohort_cs286_lung.2024-11-20.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_follow_up</th>\n",
       "      <th>survival_time</th>\n",
       "      <th>event</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>e7e95597-ef88-4442-8c01-8d633822e7b5</td>\n",
       "      <td>TCGA-XF-AAN0</td>\n",
       "      <td>Transitional Cell Papillomas and Carcinomas</td>\n",
       "      <td>Bladder</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1718.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1718.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>06f936e9-5a90-40d3-b91a-713f2b4e6e11</td>\n",
       "      <td>TCGA-GV-A3QI</td>\n",
       "      <td>Transitional Cell Papillomas and Carcinomas</td>\n",
       "      <td>Bladder</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ebedab7c-dd7c-448e-8324-78c86f13be8b</td>\n",
       "      <td>TCGA-BT-A20N</td>\n",
       "      <td>Transitional Cell Papillomas and Carcinomas</td>\n",
       "      <td>Bladder</td>\n",
       "      <td>male</td>\n",
       "      <td>not reported</td>\n",
       "      <td>Dead</td>\n",
       "      <td>795.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>795.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1029514b-a32c-4c43-a440-a5ba2709f717</td>\n",
       "      <td>TCGA-ZF-AA51</td>\n",
       "      <td>Transitional Cell Papillomas and Carcinomas</td>\n",
       "      <td>Bladder</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>f3104468-bf17-434b-8a7e-bc3b253121ea</td>\n",
       "      <td>TCGA-DK-A1AF</td>\n",
       "      <td>Transitional Cell Papillomas and Carcinomas</td>\n",
       "      <td>Bladder</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id  submitter_id  \\\n",
       "0  e7e95597-ef88-4442-8c01-8d633822e7b5  TCGA-XF-AAN0   \n",
       "1  06f936e9-5a90-40d3-b91a-713f2b4e6e11  TCGA-GV-A3QI   \n",
       "2  ebedab7c-dd7c-448e-8324-78c86f13be8b  TCGA-BT-A20N   \n",
       "3  1029514b-a32c-4c43-a440-a5ba2709f717  TCGA-ZF-AA51   \n",
       "4  f3104468-bf17-434b-8a7e-bc3b253121ea  TCGA-DK-A1AF   \n",
       "\n",
       "                                  disease_type primary_site  gender  \\\n",
       "0  Transitional Cell Papillomas and Carcinomas      Bladder    male   \n",
       "1  Transitional Cell Papillomas and Carcinomas      Bladder    male   \n",
       "2  Transitional Cell Papillomas and Carcinomas      Bladder    male   \n",
       "3  Transitional Cell Papillomas and Carcinomas      Bladder  female   \n",
       "4  Transitional Cell Papillomas and Carcinomas      Bladder  female   \n",
       "\n",
       "           race vital_status  days_to_death  days_to_last_follow_up  \\\n",
       "0         white         Dead         1718.0                     NaN   \n",
       "1         white        Alive            NaN                     NaN   \n",
       "2  not reported         Dead          795.0                     NaN   \n",
       "3         white        Alive            NaN                     NaN   \n",
       "4         white        Alive            NaN                     NaN   \n",
       "\n",
       "   survival_time  event  \n",
       "0         1718.0      1  \n",
       "1            NaN      0  \n",
       "2          795.0      1  \n",
       "3            NaN      0  \n",
       "4            NaN      0  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Acinar Cell NeoplasmsBronchus and lung</th>\n",
       "      <td>22</td>\n",
       "      <td>0.136364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adenomas and AdenocarcinomasBladder</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adenomas and AdenocarcinomasBreast</th>\n",
       "      <td>3</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adenomas and AdenocarcinomasBronchus and lung</th>\n",
       "      <td>486</td>\n",
       "      <td>0.374486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adenomas and AdenocarcinomasCorpus uteri</th>\n",
       "      <td>407</td>\n",
       "      <td>0.117936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adenomas and AdenocarcinomasKidney</th>\n",
       "      <td>828</td>\n",
       "      <td>0.266908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adenomas and AdenocarcinomasPancreas</th>\n",
       "      <td>29</td>\n",
       "      <td>0.275862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adnexal and Skin Appendage NeoplasmsBreast</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Basal Cell NeoplasmsBreast</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Complex Epithelial NeoplasmsBreast</th>\n",
       "      <td>14</td>\n",
       "      <td>0.142857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cystic, Mucinous and Serous NeoplasmsBreast</th>\n",
       "      <td>16</td>\n",
       "      <td>0.125000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cystic, Mucinous and Serous NeoplasmsBronchus and lung</th>\n",
       "      <td>14</td>\n",
       "      <td>0.214286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cystic, Mucinous and Serous NeoplasmsCorpus uteri</th>\n",
       "      <td>139</td>\n",
       "      <td>0.309353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cystic, Mucinous and Serous NeoplasmsPancreas</th>\n",
       "      <td>5</td>\n",
       "      <td>0.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ductal and Lobular NeoplasmsBreast</th>\n",
       "      <td>1054</td>\n",
       "      <td>0.137571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ductal and Lobular NeoplasmsPancreas</th>\n",
       "      <td>150</td>\n",
       "      <td>0.593333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epithelial Neoplasms, NOSBladder</th>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epithelial Neoplasms, NOSBreast</th>\n",
       "      <td>5</td>\n",
       "      <td>0.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epithelial Neoplasms, NOSCorpus uteri</th>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epithelial Neoplasms, NOSPancreas</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fibroepithelial NeoplasmsBreast</th>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GliomasBrain</th>\n",
       "      <td>602</td>\n",
       "      <td>0.815615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Not ReportedBrain</th>\n",
       "      <td>5</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Not ReportedUterus, NOS</th>\n",
       "      <td>12</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Squamous Cell NeoplasmsBladder</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Squamous Cell NeoplasmsBreast</th>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Transitional Cell Papillomas and CarcinomasBladder</th>\n",
       "      <td>409</td>\n",
       "      <td>0.442543</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    count      mean\n",
       "Acinar Cell NeoplasmsBronchus and lung                 22  0.136364\n",
       "Adenomas and AdenocarcinomasBladder                     1  0.000000\n",
       "Adenomas and AdenocarcinomasBreast                      3  0.333333\n",
       "Adenomas and AdenocarcinomasBronchus and lung         486  0.374486\n",
       "Adenomas and AdenocarcinomasCorpus uteri              407  0.117936\n",
       "Adenomas and AdenocarcinomasKidney                    828  0.266908\n",
       "Adenomas and AdenocarcinomasPancreas                   29  0.275862\n",
       "Adnexal and Skin Appendage NeoplasmsBreast              1  0.000000\n",
       "Basal Cell NeoplasmsBreast                              1  0.000000\n",
       "Complex Epithelial NeoplasmsBreast                     14  0.142857\n",
       "Cystic, Mucinous and Serous NeoplasmsBreast            16  0.125000\n",
       "Cystic, Mucinous and Serous NeoplasmsBronchus a...     14  0.214286\n",
       "Cystic, Mucinous and Serous NeoplasmsCorpus uteri     139  0.309353\n",
       "Cystic, Mucinous and Serous NeoplasmsPancreas           5  0.600000\n",
       "Ductal and Lobular NeoplasmsBreast                   1054  0.137571\n",
       "Ductal and Lobular NeoplasmsPancreas                  150  0.593333\n",
       "Epithelial Neoplasms, NOSBladder                        1  1.000000\n",
       "Epithelial Neoplasms, NOSBreast                         5  0.400000\n",
       "Epithelial Neoplasms, NOSCorpus uteri                   2  0.000000\n",
       "Epithelial Neoplasms, NOSPancreas                       1  0.000000\n",
       "Fibroepithelial NeoplasmsBreast                         2  0.000000\n",
       "GliomasBrain                                          602  0.815615\n",
       "Not ReportedBrain                                       5  0.000000\n",
       "Not ReportedUterus, NOS                                12  0.000000\n",
       "Squamous Cell NeoplasmsBladder                          1  0.000000\n",
       "Squamous Cell NeoplasmsBreast                           2  0.000000\n",
       "Transitional Cell Papillomas and CarcinomasBladder    409  0.442543"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(df.disease_type + df.primary_site).event.agg(['count', 'mean'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "submitter_id\n",
       "TCGA    4212\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.submitter_id.str[:4].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "primary_site\n",
       "Breast               0.260684\n",
       "Kidney               0.196581\n",
       "Brain                0.144112\n",
       "Corpus uteri         0.130104\n",
       "Bronchus and lung    0.123932\n",
       "Bladder              0.097816\n",
       "Pancreas             0.043922\n",
       "Uterus, NOS          0.002849\n",
       "Name: proportion, dtype: float64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.primary_site.value_counts(normalize=True, dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gender\n",
       "female    0.617759\n",
       "male      0.376306\n",
       "NaN       0.005935\n",
       "Name: proportion, dtype: float64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.gender.value_counts(normalize=True, dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "race\n",
       "white                                        0.757835\n",
       "black or african american                    0.128917\n",
       "not reported                                 0.062678\n",
       "asian                                        0.040598\n",
       "NaN                                          0.005935\n",
       "native hawaiian or other pacific islander    0.002137\n",
       "american indian or alaska native             0.001899\n",
       "Name: proportion, dtype: float64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.race.value_counts(normalize=True, dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "vital_status\n",
       "Alive           0.654796\n",
       "Dead            0.338319\n",
       "NaN             0.005935\n",
       "Not Reported    0.000950\n",
       "Name: proportion, dtype: float64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.vital_status.value_counts(normalize=True, dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "days_to_death\n",
       "NaN       0.663580\n",
       "0.0       0.001899\n",
       "454.0     0.001425\n",
       "385.0     0.001425\n",
       "321.0     0.001187\n",
       "            ...   \n",
       "584.0     0.000237\n",
       "3926.0    0.000237\n",
       "2483.0    0.000237\n",
       "3873.0    0.000237\n",
       "456.0     0.000237\n",
       "Name: proportion, Length: 930, dtype: float64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.days_to_death.value_counts(normalize=True, dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "days_to_last_follow_up\n",
       "NaN    1.0\n",
       "Name: proportion, dtype: float64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.days_to_last_follow_up.value_counts(normalize=True, dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "survival_time\n",
       "NaN       0.663580\n",
       "0.0       0.001899\n",
       "454.0     0.001425\n",
       "385.0     0.001425\n",
       "321.0     0.001187\n",
       "            ...   \n",
       "584.0     0.000237\n",
       "3926.0    0.000237\n",
       "2483.0    0.000237\n",
       "3873.0    0.000237\n",
       "456.0     0.000237\n",
       "Name: proportion, Length: 930, dtype: float64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.survival_time.value_counts(normalize=True, dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "event\n",
       "0    0.661681\n",
       "1    0.338319\n",
       "Name: proportion, dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.event.value_counts(normalize=True, dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Request optim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "cases = data_utils.pull_cases()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': '32d9935a-f5cd-43b3-88bd-09e6e2b9090d',\n",
       " 'primary_site': 'Bronchus and lung',\n",
       " 'disease_type': 'Adenomas and Adenocarcinomas',\n",
       " 'case_id': '32d9935a-f5cd-43b3-88bd-09e6e2b9090d',\n",
       " 'diagnoses': [{'ajcc_pathologic_t': 'T1b',\n",
       "   'days_to_last_follow_up': 689,\n",
       "   'days_to_recurrence': None,\n",
       "   'age_at_diagnosis': 23341,\n",
       "   'ajcc_pathologic_stage': 'Stage IIA',\n",
       "   'updated_datetime': '2023-10-06T12:23:18.779951-05:00',\n",
       "   'ajcc_pathologic_n': 'N1',\n",
       "   'ajcc_pathologic_m': 'M0',\n",
       "   'diagnosis_id': 'b40b0b2e-be0a-5a95-a43a-518b8311f611',\n",
       "   'days_to_last_known_disease_status': None,\n",
       "   'treatments': [{'days_to_treatment_end': None,\n",
       "     'days_to_treatment_start': None,\n",
       "     'treatment_id': 'd64775f4-012b-5dd4-ab83-14faa6f08dee',\n",
       "     'treatment_type': 'Pharmaceutical Therapy, NOS',\n",
       "     'therapeutic_agents': None},\n",
       "    {'treatment_id': 'f4f5773c-c84a-5424-9f37-be733d29a815',\n",
       "     'treatment_type': 'Radiation Therapy, NOS',\n",
       "     'therapeutic_agents': None}],\n",
       "   'last_known_disease_status': 'not reported'}],\n",
       " 'demographic': {'race': 'white',\n",
       "  'gender': 'female',\n",
       "  'vital_status': 'Alive',\n",
       "  'year_of_death': None,\n",
       "  'year_of_birth': 1949}}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cases[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|\u001b[32m██████████\u001b[0m| 585/585 [00:00<00:00, 119341.82cases/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping case 9b714a42-62e8-4b33-947e-6c4850725afd because it has no demographic data\n",
      "Skipping case a16d7bf7-07b7-4320-969c-fc68dd564a92 because it has no demographic data\n",
      "Skipping case 4b5ff30e-65fc-4434-b174-9d699594c658 because it has no demographic data\n",
      "Skipping case abb86d09-d4ff-41f4-ba22-185dc48347c9 because it has no demographic data\n",
      "Skipping case 56027785-fb10-492c-b021-ea9e3a1979db because it has no demographic data\n",
      "Skipping case aebd0313-23be-46a8-abc6-b16c531c3a8e because it has no demographic data\n",
      "Skipping case b3c31fdc-26a7-4b39-8127-8006307eb1f5 because it has no demographic data\n",
      "Skipping case 61cd3ef8-eee4-4266-b01d-09338c89740a because it has no demographic data\n",
      "Skipping case 6313c7cb-5631-42b2-93e2-592f86e84551 because it has no demographic data\n",
      "Skipping case 631dbd88-fc3e-4ed8-b4a7-2da137a49a11 because it has no demographic data\n",
      "Skipping case 654affac-f8ed-40bb-9f9f-a293227b06a6 because it has no demographic data\n",
      "Skipping case 67a2593d-b986-4f84-a51c-7f2fb12e54ad because it has no demographic data\n",
      "Skipping case 67b5fb5d-2484-4aa5-aa7c-c6c5ecbc91b2 because it has no demographic data\n",
      "Skipping case c43c8eea-7014-4abe-b805-397a9062d3e0 because it has no demographic data\n",
      "Skipping case 77bb204d-4e68-4c93-905f-9ef4c53b1ce2 because it has no demographic data\n",
      "Skipping case cf411fc0-0cce-42d2-ae79-24de1a105c05 because it has no demographic data\n",
      "Skipping case 79684460-f74c-4954-80fa-b418827bf2ef because it has no demographic data\n",
      "Skipping case 3136b041-24b9-42c5-b4a8-70f04c63a177 because it has no demographic data\n",
      "Skipping case 3650e5b2-ef58-4b14-97b2-8698f2c160f8 because it has no demographic data\n",
      "Skipping case 81b71cff-da58-47cf-ad00-0a42489a7a0b because it has no demographic data\n",
      "Skipping case 392ca7f3-d52e-4aa7-a9ba-4c7d9aa6401c because it has no demographic data\n",
      "Skipping case 87eeaf49-3c02-402f-9ae4-3627391e7749 because it has no demographic data\n",
      "Skipping case 3d0cbd55-4eda-4c52-93cc-a74381ea7c59 because it has no demographic data\n",
      "Skipping case 4432a02d-91de-4a7d-875f-684865f3c95d because it has no demographic data\n",
      "Skipping case 4999df33-3a47-4feb-a264-c5d71b271d8c because it has no demographic data\n",
      "Skipping case a19d84a1-b316-44d3-90b1-0d14b0fd7a14 because it has no demographic data\n",
      "Skipping case a5d0c704-2888-4cff-805e-4bbe4636117c because it has no demographic data\n",
      "Skipping case 51e49548-1ad0-4c83-ae8a-2fa95bf4712a because it has no demographic data\n",
      "Skipping case a8f3da11-890c-4fff-bfc5-d99357dc4e46 because it has no demographic data\n",
      "Skipping case 5730b09c-a69f-425a-8af9-61172f9dfe56 because it has no demographic data\n",
      "Skipping case b2c6939b-d8ec-4a07-8ea6-06c4553ee446 because it has no demographic data\n",
      "Skipping case b5cdf68a-9903-4f82-9017-ab88d2b73d4f because it has no demographic data\n",
      "Skipping case 62633a11-db23-481f-add3-853b2acac1bf because it has no demographic data\n",
      "Skipping case bd693ad9-6ce9-469a-a3f9-c2779c32d6bc because it has no demographic data\n",
      "Skipping case c05b1a5c-2f6d-448a-94c8-58cce3d34b37 because it has no demographic data\n",
      "Skipping case 733087d2-98ed-4ac7-8b48-98f0fe1c4c0f because it has no demographic data\n",
      "Skipping case d020e685-0435-4b0a-b03f-45853e1c6eed because it has no demographic data\n",
      "Skipping case d6a08727-affd-4a59-837b-ea71741d7f09 because it has no demographic data\n",
      "Skipping case dcf8e752-e95c-4087-a0f5-c041787c8569 because it has no demographic data\n",
      "Skipping case e44e279b-103e-436e-8988-1139b000bdfd because it has no demographic data\n",
      "Skipping case e998a607-ac73-4849-afaa-57eee49cf051 because it has no demographic data\n",
      "Skipping case edd05f61-35f8-4133-a579-f9f32dff3bb8 because it has no demographic data\n",
      "Skipping case f29b5aef-d53e-457d-b140-b0b5fd1960bc because it has no demographic data\n",
      "Skipping case f4b83449-7d0e-49f3-be53-81cf0d485ed8 because it has no demographic data\n",
      "Skipping case f86e1da7-bad6-4eb8-ae4c-fa9421308d47 because it has no demographic data\n",
      "Skipping case fae64c9d-5204-4dc6-a561-bf59506aabcc because it has no demographic data\n",
      "Skipping case fddc4d29-1ae8-451d-a13e-8ef6492be201 because it has no demographic data\n",
      "Skipping case c390b6ad-14c8-49ea-9b38-798c9fad5b62 because it has no demographic data\n",
      "Skipping case 0147edcf-0ca5-4136-bd01-43d05e2843f4 because it has no demographic data\n",
      "Skipping case 07119baf-64a7-454c-b1b0-c769b506a63d because it has no demographic data\n",
      "Skipping case 0d96def2-cf0a-4659-94e5-f08744c57cc6 because it has no demographic data\n",
      "Skipping case 175cb088-9b10-4fa9-95c0-7837fdc7bcf9 because it has no demographic data\n",
      "Skipping case 19ca36e6-2154-4224-89b1-117a4a4407f6 because it has no demographic data\n",
      "Skipping case 257c9416-e86a-4c71-9dee-7212c2e56351 because it has no demographic data\n",
      "Skipping case 25c9f984-1a5a-4cca-9ad1-39f759ea7174 because it has no demographic data\n",
      "Skipping case 2dca3588-51e1-48f7-8c9b-7b7e53c3fb33 because it has no demographic data\n",
      "Skipping case 3037e91e-213e-43e3-92e4-2fcd68988bde because it has no demographic data\n",
      "Skipping case 8b3a111a-aa46-41b9-bbec-28f4290da939 because it has no demographic data\n",
      "Skipping case 94ee639e-3cb4-4943-b5f1-a89eb902153c because it has no demographic data\n",
      "Skipping case 494d50b6-578e-441b-8195-4b4d26c0d810 because it has no demographic data\n",
      "Skipping case a56d2b48-375d-4193-91a9-f589c39523fb because it has no demographic data\n",
      "Skipping case 53d7b0c2-b8ed-43c8-9ffd-a586e0a68086 because it has no demographic data\n",
      "Skipping case abf6eb5f-fa06-4032-8cff-c8522b0a3c9a because it has no demographic data\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "df = data_utils.process_cases(cases)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>year_of_birth</th>\n",
       "      <th>year_of_death</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>num_diagnoses</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>days_to_diagnosis</th>\n",
       "      <th>ajcc_pathologic_stage</th>\n",
       "      <th>days_to_last_follow_up</th>\n",
       "      <th>days_to_recurrence</th>\n",
       "      <th>days_to_last_known_disease_status</th>\n",
       "      <th>survival_time</th>\n",
       "      <th>event</th>\n",
       "      <th>ajcc_pathologic_stage_coarse</th>\n",
       "      <th>age_at_diagnosis_years</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>32d9935a-f5cd-43b3-88bd-09e6e2b9090d</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Bronchus and lung</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>1949.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>23341.0</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>689.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>689.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>63.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7c7d777b-bf00-4c24-97fc-922580d5605f</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Bronchus and lung</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>1944.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>24738.0</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage IA</td>\n",
       "      <td>603.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>603.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7ca711ec-9ac0-4fa7-8a60-32ae84b8789a</td>\n",
       "      <td>Acinar Cell Neoplasms</td>\n",
       "      <td>Bronchus and lung</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>1949.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>23201.0</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>1071.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1071.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>63.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>349e6f38-2c67-4a69-a777-b9173a2a27cf</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Bronchus and lung</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>1921.0</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>987.0</td>\n",
       "      <td>1</td>\n",
       "      <td>31234.0</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage IB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>987.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>85.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>35cb7841-9b09-465a-90c5-e3b8a9faad49</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Bronchus and lung</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>1925.0</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>995.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>Stage IIB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>995.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id                  disease_type  \\\n",
       "0  32d9935a-f5cd-43b3-88bd-09e6e2b9090d  Adenomas and Adenocarcinomas   \n",
       "1  7c7d777b-bf00-4c24-97fc-922580d5605f  Adenomas and Adenocarcinomas   \n",
       "2  7ca711ec-9ac0-4fa7-8a60-32ae84b8789a         Acinar Cell Neoplasms   \n",
       "3  349e6f38-2c67-4a69-a777-b9173a2a27cf  Adenomas and Adenocarcinomas   \n",
       "4  35cb7841-9b09-465a-90c5-e3b8a9faad49  Adenomas and Adenocarcinomas   \n",
       "\n",
       "        primary_site  gender   race  year_of_birth  year_of_death  \\\n",
       "0  Bronchus and lung  female  white         1949.0            NaN   \n",
       "1  Bronchus and lung    male  white         1944.0            NaN   \n",
       "2  Bronchus and lung    male  white         1949.0            NaN   \n",
       "3  Bronchus and lung  female  white         1921.0         2008.0   \n",
       "4  Bronchus and lung  female  white         1925.0         2006.0   \n",
       "\n",
       "  vital_status  days_to_death  num_diagnoses  age_at_diagnosis  \\\n",
       "0        Alive            NaN              1           23341.0   \n",
       "1        Alive            NaN              1           24738.0   \n",
       "2        Alive            NaN              1           23201.0   \n",
       "3         Dead          987.0              1           31234.0   \n",
       "4         Dead          995.0              1               NaN   \n",
       "\n",
       "  days_to_diagnosis ajcc_pathologic_stage  days_to_last_follow_up  \\\n",
       "0              None             Stage IIA                   689.0   \n",
       "1              None              Stage IA                   603.0   \n",
       "2              None             Stage IIA                  1071.0   \n",
       "3              None              Stage IB                     NaN   \n",
       "4              None             Stage IIB                     NaN   \n",
       "\n",
       "  days_to_recurrence days_to_last_known_disease_status  survival_time  event  \\\n",
       "0               None                              None          689.0      0   \n",
       "1               None                              None          603.0      0   \n",
       "2               None                              None         1071.0      0   \n",
       "3               None                              None          987.0      1   \n",
       "4               None                              None          995.0      1   \n",
       "\n",
       "  ajcc_pathologic_stage_coarse  age_at_diagnosis_years  \n",
       "0                     Stage II                    63.0  \n",
       "1                      Stage I                    67.0  \n",
       "2                     Stage II                    63.0  \n",
       "3                      Stage I                    85.0  \n",
       "4                     Stage II                     NaN  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "event\n",
       "0    334\n",
       "1    188\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.event.value_counts(normalize=False, dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ajcc_pathologic_stage_coarse\n",
       "Stage I      0.247312\n",
       "Stage II     0.435484\n",
       "Stage III    0.552941\n",
       "Stage IV     0.615385\n",
       "Name: event, dtype: float64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby('ajcc_pathologic_stage_coarse').event.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('data/case_data.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/08/91_vkqq52qg5phplpcj85z480000gn/T/ipykernel_17824/4263094649.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  df.groupby(pd.qcut(df.age_at_diagnosis_years, 5)).event.mean()\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "age_at_diagnosis_years\n",
       "(32.999, 57.0]    0.382353\n",
       "(57.0, 63.0]      0.313725\n",
       "(63.0, 69.0]      0.313131\n",
       "(69.0, 74.0]      0.354167\n",
       "(74.0, 88.0]      0.423913\n",
       "Name: event, dtype: float64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(pd.qcut(df.age_at_diagnosis_years, 5)).event.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "survival_time\n",
       "NaN       9\n",
       "0.0       4\n",
       "791.0     4\n",
       "568.0     3\n",
       "626.0     3\n",
       "         ..\n",
       "1974.0    1\n",
       "592.0     1\n",
       "1621.0    1\n",
       "455.0     1\n",
       "922.0     1\n",
       "Name: count, Length: 438, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.survival_time.value_counts(normalize=False, dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'race': 'white',\n",
       " 'gender': 'male',\n",
       " 'vital_status': 'Dead',\n",
       " 'days_to_death': 1293,\n",
       " 'year_of_death': 2007,\n",
       " 'year_of_birth': 1943}"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cases[9]['demographic']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'days_to_last_follow_up': 689,\n",
       "  'days_to_recurrence': None,\n",
       "  'age_at_diagnosis': 23341,\n",
       "  'days_to_last_known_disease_status': None,\n",
       "  'treatments': [{'treatment_id': 'd64775f4-012b-5dd4-ab83-14faa6f08dee',\n",
       "    'therapeutic_agents': None,\n",
       "    'treatment_or_therapy': 'yes',\n",
       "    'created_datetime': '2019-04-28T12:54:29.887722-05:00'},\n",
       "   {'treatment_id': 'f4f5773c-c84a-5424-9f37-be733d29a815',\n",
       "    'therapeutic_agents': None,\n",
       "    'treatment_or_therapy': 'no',\n",
       "    'created_datetime': None}],\n",
       "  'last_known_disease_status': 'not reported'}]"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cases[0]['diagnoses']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'data_format': 'SVS',\n",
       "  'access': 'open',\n",
       "  'file_id': '6f09ba8e-653d-48a4-bea5-bc93965d570c',\n",
       "  'data_type': 'Slide Image'},\n",
       " {'data_format': 'SVS',\n",
       "  'access': 'open',\n",
       "  'file_id': 'c8f6bd03-0a77-4c93-8caf-fef4b1569085',\n",
       "  'data_type': 'Slide Image'},\n",
       " {'data_format': 'SVS',\n",
       "  'access': 'open',\n",
       "  'file_id': '689cd91c-6a4f-4423-84e5-3adbb9829a99',\n",
       "  'data_type': 'Slide Image'}]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[f for f in cases[8]['files'] if f['data_type'] == 'Slide Image']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'data_format': 'BAM',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '05ce3de9-6246-488a-ae6d-68061bec96a6',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'analysis': {'created_datetime': '2016-05-30T15:49:12.065364-05:00'}},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '8e7fa01e-aff4-486f-a95f-7aeecd005270',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'BAM',\n",
       "  'access': 'controlled',\n",
       "  'file_id': 'a427d468-f261-4d1f-8d15-a18b0037a908',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'analysis': {'created_datetime': '2021-12-13T16:41:23.239662-06:00'}},\n",
       " {'data_format': 'SVS',\n",
       "  'access': 'open',\n",
       "  'file_id': '6f09ba8e-653d-48a4-bea5-bc93965d570c',\n",
       "  'data_type': 'Slide Image'},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '7fc41a9c-65f8-425a-8a93-5cbce85b6353',\n",
       "  'data_type': 'Simple Germline Variation',\n",
       "  'analysis': {'created_datetime': '2023-02-19T04:18:49.453628-06:00'}},\n",
       " {'data_format': 'TXT',\n",
       "  'access': 'open',\n",
       "  'file_id': 'cbc5164d-0f09-4ec9-b1b2-a741c1fe6d6f',\n",
       "  'data_type': 'Copy Number Segment',\n",
       "  'analysis': {'created_datetime': '2016-05-03T03:50:15.585536-05:00'}},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '204f3cf1-8dfc-4aea-92fc-b345a59b8834',\n",
       "  'data_type': 'Simple Germline Variation',\n",
       "  'analysis': {'created_datetime': '2023-02-19T04:18:49.453628-06:00'}},\n",
       " {'data_format': 'VCF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '0d02bb8b-c3a6-4b71-8aad-fcb04019a0d3',\n",
       "  'data_type': 'Raw Simple Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2016-06-03T17:26:59.908252-05:00'}},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'open',\n",
       "  'file_id': '44356190-4c0e-497e-a7f1-07ba2f177274',\n",
       "  'data_type': 'Gene Level Copy Number',\n",
       "  'analysis': {'created_datetime': '2023-04-26T14:32:44.496393-05:00'}},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': 'c0cb0e6a-b19e-46a5-b750-91ff91385071',\n",
       "  'data_type': 'Clinical Supplement'},\n",
       " {'data_format': 'IDAT',\n",
       "  'access': 'open',\n",
       "  'file_id': '979e5135-98a1-4b89-b394-1b48abf61d20',\n",
       "  'data_type': 'Masked Intensities',\n",
       "  'analysis': {'created_datetime': '2022-02-15T16:35:31.777248-06:00'}},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '62e98ea1-ff2c-402d-acb7-604ae5648ca0',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'VCF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '806ff8a2-3760-4fce-8479-6913064ac584',\n",
       "  'data_type': 'Raw Simple Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2016-05-08T13:25:04.942315-05:00'}},\n",
       " {'data_format': 'SVS',\n",
       "  'access': 'open',\n",
       "  'file_id': 'c8f6bd03-0a77-4c93-8caf-fef4b1569085',\n",
       "  'data_type': 'Slide Image'},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '0277f676-24c7-4a4b-90d3-3897e3b27a27',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'TXT',\n",
       "  'access': 'open',\n",
       "  'file_id': '9aea5b00-1843-4a8c-9f5d-72e01bc1b5ce',\n",
       "  'data_type': 'Copy Number Segment',\n",
       "  'analysis': {'created_datetime': '2016-05-02T23:39:48.723868-05:00'}},\n",
       " {'data_format': 'IDAT',\n",
       "  'access': 'open',\n",
       "  'file_id': '11012405-2928-4e26-97b6-ac8c16e3b920',\n",
       "  'data_type': 'Masked Intensities',\n",
       "  'analysis': {'created_datetime': '2022-02-15T16:35:31.777248-06:00'}},\n",
       " {'data_format': 'TXT',\n",
       "  'access': 'open',\n",
       "  'file_id': '0b42cdbb-4484-4850-aa42-10bd2581552b',\n",
       "  'data_type': 'Masked Copy Number Segment',\n",
       "  'analysis': {'created_datetime': '2016-05-03T03:50:15.585536-05:00'}},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '8d3ba9f4-0945-4937-9d17-bf30c4934331',\n",
       "  'data_type': 'Transcript Fusion',\n",
       "  'analysis': {'created_datetime': '2021-12-13T20:37:59.571785-06:00'}},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': 'f4b19626-a243-43a6-8257-cfe44cefa118',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': 'cd9bb8a0-bb51-4319-b172-ec35c8c5e0e4',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'BCR XML',\n",
       "  'access': 'open',\n",
       "  'file_id': '9e81fc11-09a0-4249-a92f-c0d62e9457cb',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'BAM',\n",
       "  'access': 'controlled',\n",
       "  'file_id': 'c3460c92-3108-4e57-aabb-147b7ceb0541',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'analysis': {'created_datetime': '2021-12-13T17:20:43.438166-06:00'}},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '5e339e00-02bd-4d98-b7b4-d06f4036fea0',\n",
       "  'data_type': 'Splice Junction Quantification',\n",
       "  'analysis': {'created_datetime': '2021-12-13T19:08:34.262706-06:00'}},\n",
       " {'data_format': 'BAM',\n",
       "  'access': 'controlled',\n",
       "  'file_id': 'a7b56495-9e9a-4317-9dbb-b7c3574d5712',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'analysis': {'created_datetime': '2016-05-30T11:58:58.664495-05:00'}},\n",
       " {'data_format': 'BEDPE',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '2e3a376e-c619-4028-bee0-a716b1b851ad',\n",
       "  'data_type': 'Transcript Fusion',\n",
       "  'analysis': {'created_datetime': '2021-12-13T20:10:04.427642-06:00'}},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': 'fb04acf9-524e-4bef-87c2-bf5a71713557',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '42bf5eb2-bc49-45be-b18a-290f712b006c',\n",
       "  'data_type': 'Clinical Supplement'},\n",
       " {'data_format': 'BAM',\n",
       "  'access': 'controlled',\n",
       "  'file_id': 'c92a7a64-c0eb-4e09-917f-e8cd8eb3aa4f',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'analysis': {'created_datetime': '2022-02-03T13:02:38.513347-06:00'}},\n",
       " {'data_format': 'BEDPE',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '76858df8-f001-402b-ae63-9646f5abee19',\n",
       "  'data_type': 'Transcript Fusion',\n",
       "  'analysis': {'created_datetime': '2021-12-13T20:10:02.872386-06:00'}},\n",
       " {'data_format': 'TXT',\n",
       "  'access': 'open',\n",
       "  'file_id': '3387815f-4667-4bcf-95b8-8679583f7fa1',\n",
       "  'data_type': 'Allele-specific Copy Number Segment',\n",
       "  'analysis': {'created_datetime': '2023-04-25T17:30:12.860163-05:00'}},\n",
       " {'data_format': 'VCF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': 'fb7b5710-1607-41c4-a23c-23b488f4a427',\n",
       "  'data_type': 'Raw Simple Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2016-05-08T13:25:03.727930-05:00'}},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': 'c9b06314-d440-45f7-adc0-acf69b55b0f3',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '24f6f4f9-c9b0-4a26-836c-5c85bcb10eda',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'MAF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '3ff734ed-7762-491a-b64a-185d907770fc',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2016-06-03T17:26:59.908252-05:00'}},\n",
       " {'data_format': 'VCF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '57e32246-c795-440d-9b08-ade7a0b9b805',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2016-05-08T13:25:04.942315-05:00'}},\n",
       " {'data_format': 'MAF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': 'e0eb1975-f36c-4077-8b8e-a302995fdb65',\n",
       "  'data_type': 'Aggregated Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2022-02-04T00:39:58.032418-06:00'}},\n",
       " {'data_format': 'BEDPE',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '3d0473b1-c90b-45f8-a790-6a896945bbb5',\n",
       "  'data_type': 'Transcript Fusion',\n",
       "  'analysis': {'created_datetime': '2021-12-13T20:37:59.571785-06:00'}},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'open',\n",
       "  'file_id': 'dfad0503-0ef3-4599-991d-f1dcd2e87d97',\n",
       "  'data_type': 'Gene Level Copy Number',\n",
       "  'analysis': {'created_datetime': '2020-03-11T17:02:05.151693-05:00'}},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'open',\n",
       "  'file_id': 'ecca4e4c-70fe-416a-94cc-fe20dd963284',\n",
       "  'data_type': 'Gene Level Copy Number',\n",
       "  'analysis': {'created_datetime': '2023-04-25T17:30:12.860163-05:00'}},\n",
       " {'data_format': 'VCF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '96e9485b-e1ab-415d-a83f-1065b8744770',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2016-05-08T13:25:03.727930-05:00'}},\n",
       " {'data_format': 'BCR SSF XML',\n",
       "  'access': 'open',\n",
       "  'file_id': 'acdf1ec6-0af8-4f6a-9dce-e78b0a8efc27',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': 'd45f35c0-9d3b-42ce-b0ba-6e53fe266014',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'BEDPE',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '54fcc54b-fd96-4c5f-9ed1-004462110bc1',\n",
       "  'data_type': 'Transcript Fusion',\n",
       "  'analysis': {'created_datetime': '2021-12-13T20:37:57.733637-06:00'}},\n",
       " {'data_format': 'CEL',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '42217378-fc3b-4fe8-baf8-710a5a2636d9',\n",
       "  'data_type': 'Raw Intensities'},\n",
       " {'data_format': 'CEL',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '0c4ca3a6-6223-4f66-b0b3-bced4dee0aab',\n",
       "  'data_type': 'Raw Intensities'},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '59015dfd-045b-4a80-9b4a-3cc3a41707f0',\n",
       "  'data_type': 'Clinical Supplement'},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'open',\n",
       "  'file_id': 'd107044e-70d2-43e0-b567-dfcc36d0c26d',\n",
       "  'data_type': 'Protein Expression Quantification'},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'controlled',\n",
       "  'file_id': 'd801f54d-36b0-4d61-bed9-ae77b3b54655',\n",
       "  'data_type': 'Splice Junction Quantification',\n",
       "  'analysis': {'created_datetime': '2021-12-13T19:21:54.686714-06:00'}},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '29c7eda9-f523-48bc-ac38-86d0b476ac44',\n",
       "  'data_type': 'Transcript Fusion',\n",
       "  'analysis': {'created_datetime': '2021-12-13T20:37:57.733637-06:00'}},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '27bafc9d-b95c-4b49-b3e3-1c90bb1c15dd',\n",
       "  'data_type': 'Clinical Supplement'},\n",
       " {'data_format': 'BCR XML',\n",
       "  'access': 'open',\n",
       "  'file_id': '82f544b7-18df-4927-9140-fef2a29deedf',\n",
       "  'data_type': 'Clinical Supplement'},\n",
       " {'data_format': 'MAF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': 'f9b1d124-127f-4dc3-b108-c7015dcdae64',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2016-05-08T13:25:04.942315-05:00'}},\n",
       " {'data_format': 'SVS',\n",
       "  'access': 'open',\n",
       "  'file_id': '689cd91c-6a4f-4423-84e5-3adbb9829a99',\n",
       "  'data_type': 'Slide Image'},\n",
       " {'data_format': 'TXT',\n",
       "  'access': 'open',\n",
       "  'file_id': '4e5f2919-46f1-42be-a085-7a88ea1d77d0',\n",
       "  'data_type': 'Masked Copy Number Segment',\n",
       "  'analysis': {'created_datetime': '2016-05-02T23:39:48.723868-05:00'}},\n",
       " {'data_format': 'BAM',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '91c8e80e-d323-4c35-a627-bb3ced73877a',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'analysis': {'created_datetime': '2021-12-13T16:41:26.035850-06:00'}},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '6f77c7b9-e114-4305-a12a-7ca8c0c35a88',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'VCF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '7ed49a1f-ecff-443f-ae92-d6f94e3e8a42',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2022-02-03T19:53:23.197170-06:00'}},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': 'b8b73477-ad95-4e9d-a758-cc536754d420',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'controlled',\n",
       "  'file_id': 'd0001212-66af-4267-a361-e6f35d46eac1',\n",
       "  'data_type': 'Transcript Fusion',\n",
       "  'analysis': {'created_datetime': '2021-12-13T20:10:04.427642-06:00'}},\n",
       " {'data_format': 'BAM',\n",
       "  'access': 'controlled',\n",
       "  'file_id': 'b5ada179-1b87-4cd7-aa63-448022e81cc1',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'analysis': {'created_datetime': '2022-02-03T13:01:50.294112-06:00'}},\n",
       " {'data_format': 'TXT',\n",
       "  'access': 'open',\n",
       "  'file_id': 'f888ded9-21bd-441f-bf16-4564889b7516',\n",
       "  'data_type': 'Methylation Beta Value',\n",
       "  'analysis': {'created_datetime': '2022-02-15T16:35:31.777248-06:00'}},\n",
       " {'data_format': 'MAF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': 'acffaca7-3a54-4e7d-8b43-2521aa7f49b4',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2016-05-08T13:25:03.727930-05:00'}},\n",
       " {'data_format': 'TXT',\n",
       "  'access': 'open',\n",
       "  'file_id': 'c92a3315-fd2d-4c79-904c-d23ae12e44db',\n",
       "  'data_type': 'Allele-specific Copy Number Segment',\n",
       "  'analysis': {'created_datetime': '2020-03-11T17:02:05.151693-05:00'}},\n",
       " {'data_format': 'MAF',\n",
       "  'access': 'open',\n",
       "  'file_id': '63b1b642-8519-454d-b81b-5a2ce519e96b',\n",
       "  'data_type': 'Masked Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2022-02-04T00:39:58.032418-06:00'}},\n",
       " {'data_format': 'VCF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '94a44934-578d-452d-bcd3-0586336414d1',\n",
       "  'data_type': 'Raw Simple Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2022-02-03T15:45:47.828285-06:00'}},\n",
       " {'data_format': 'TXT',\n",
       "  'access': 'open',\n",
       "  'file_id': '9d7e7025-21d6-48c5-8067-40c37fc850d0',\n",
       "  'data_type': 'Isoform Expression Quantification',\n",
       "  'analysis': {'created_datetime': '2018-03-20T07:37:15.602852-05:00'}},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '70a0b394-9926-4ea9-baed-df5c5b270276',\n",
       "  'data_type': 'Transcript Fusion',\n",
       "  'analysis': {'created_datetime': '2021-12-13T20:10:02.872386-06:00'}},\n",
       " {'data_format': 'VCF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '08b71509-8f52-420a-9a31-fc6e5d8bb21b',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2016-06-03T17:26:59.908252-05:00'}},\n",
       " {'data_format': 'BAM',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '75e4fad3-d956-4728-838b-95ea18532b6d',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'analysis': {'created_datetime': '2021-12-13T17:20:40.617859-06:00'}},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '7ee8414d-c1e1-4c42-935e-ab397ea7e968',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': 'a327a35d-246c-4df6-9213-7f01353f4669',\n",
       "  'data_type': 'Clinical Supplement'},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '015dbbda-030b-488e-942d-b56ccd6c965e',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'open',\n",
       "  'file_id': 'f700b996-2ccf-4f59-9249-da7f087054ba',\n",
       "  'data_type': 'Gene Expression Quantification',\n",
       "  'analysis': {'created_datetime': '2021-12-13T19:08:34.262706-06:00'}},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '5af9208f-727f-42ef-a76a-47a577ba1d43',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'MAF',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '1df8ee31-a243-4f55-9769-173498c8a33d',\n",
       "  'data_type': 'Annotated Somatic Mutation',\n",
       "  'analysis': {'created_datetime': '2022-02-03T19:53:23.197170-06:00'}},\n",
       " {'data_format': 'TXT',\n",
       "  'access': 'open',\n",
       "  'file_id': 'a51d1524-3daa-4f38-b9ea-d705bf96f08e',\n",
       "  'data_type': 'miRNA Expression Quantification',\n",
       "  'analysis': {'created_datetime': '2018-03-20T07:37:15.602852-05:00'}},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '75a6bfbc-2ca3-4115-9fe2-f9dd8a50d68b',\n",
       "  'data_type': 'Clinical Supplement'},\n",
       " {'data_format': 'TSV',\n",
       "  'access': 'open',\n",
       "  'file_id': '0e9c2659-4131-4274-b0b4-3e191239f5f3',\n",
       "  'data_type': 'Gene Expression Quantification',\n",
       "  'analysis': {'created_datetime': '2021-12-13T19:21:54.686714-06:00'}},\n",
       " {'data_format': 'BAM',\n",
       "  'access': 'controlled',\n",
       "  'file_id': '8be401e7-1fa3-4ae3-9f18-660941824671',\n",
       "  'data_type': 'Aligned Reads',\n",
       "  'analysis': {'created_datetime': '2016-05-30T12:00:15.240098-05:00'}},\n",
       " {'data_format': 'PDF',\n",
       "  'access': 'open',\n",
       "  'file_id': 'c597cced-135b-4cb5-815c-2d5dded6ae02',\n",
       "  'data_type': 'Pathology Report'},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': '88d5e643-0514-4470-a09d-4da22e83a84b',\n",
       "  'data_type': 'Biospecimen Supplement'},\n",
       " {'data_format': 'BCR Biotab',\n",
       "  'access': 'open',\n",
       "  'file_id': 'b32570bc-1922-4571-b2c0-a38bfdd57c4b',\n",
       "  'data_type': 'Biospecimen Supplement'}]"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cases[8]['files']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get data for single case id\n",
    "\n",
    "1. Use `files` endpoint to pull file_ids of the following `data_type`'s:\n",
    "    1. WSI\n",
    "    2. Gene Expression Quantification\n",
    "    3. Gene Level copy number\n",
    "2. Use `data` endpoint to download files corresponding to `file_id`'s pulled in step 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_CASE_ID = 'e7e95597-ef88-4442-8c01-8d633822e7b5'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_case_metadata(case_id):\n",
    "    url = \"https://api.gdc.cancer.gov/files\"\n",
    "    \n",
    "    # Define the filters for the query\n",
    "    filters = {\n",
    "        \"op\": \"and\",\n",
    "        \"content\": [\n",
    "            {\n",
    "                \"op\": \"in\",\n",
    "                \"content\": {\n",
    "                    \"field\": \"cases.case_id\",\n",
    "                    \"value\": [case_id]\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "                \"op\": \"in\",\n",
    "                \"content\": {\n",
    "                    \"field\": \"files.data_type\",\n",
    "                    \"value\": [\n",
    "                        \"Slide Image\",\n",
    "                        \"Gene Expression Quantification\",\n",
    "                        # \"Simple Germline Variation\", NOTE this data format is controlled\n",
    "                        \"Gene Level Copy Number\",\n",
    "                    ]\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "    \n",
    "    # Define the parameters for the request\n",
    "    params = {\n",
    "        \"filters\": filters,\n",
    "        \"fields\": \"file_id,file_name,experimental_strategy,data_type,data_format,file_size\",\n",
    "        \"format\": \"JSON\",\n",
    "        \"size\": \"1000\"\n",
    "    }\n",
    "    \n",
    "    # Make the request to the GDC API\n",
    "    response = requests.post(url, json=params)\n",
    "    \n",
    "    # Check if the request was successful\n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "        return data[\"data\"][\"hits\"]\n",
    "    else:\n",
    "        response.raise_for_status()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': '67e5f3a1-c5f8-46b5-a084-71378cd3dfd6',\n",
       "  'data_format': 'SVS',\n",
       "  'file_name': 'TCGA-XF-AAN0-01Z-00-DX1.B9D6C207-F124-4558-ABD4-2846F42C5F66.svs',\n",
       "  'file_id': '67e5f3a1-c5f8-46b5-a084-71378cd3dfd6',\n",
       "  'data_type': 'Slide Image',\n",
       "  'experimental_strategy': 'Diagnostic Slide',\n",
       "  'file_size': 1745531341},\n",
       " {'id': '5fc243af-89e9-41d4-bf2f-3deb6459bf4b',\n",
       "  'data_format': 'TSV',\n",
       "  'file_name': 'TCGA-BLCA.a2cfc450-4766-43ef-9d77-3fbb1f0f431a.gene_level_copy_number.v36.tsv',\n",
       "  'file_id': '5fc243af-89e9-41d4-bf2f-3deb6459bf4b',\n",
       "  'data_type': 'Gene Level Copy Number',\n",
       "  'experimental_strategy': 'Genotyping Array',\n",
       "  'file_size': 3446837},\n",
       " {'id': 'ac728bb8-2b0c-4a0a-9c80-23e0fb768393',\n",
       "  'data_format': 'TSV',\n",
       "  'file_name': '7fd86c81-d69b-4e3e-b60f-f4d53eb5e2a5.rna_seq.augmented_star_gene_counts.tsv',\n",
       "  'file_id': 'ac728bb8-2b0c-4a0a-9c80-23e0fb768393',\n",
       "  'data_type': 'Gene Expression Quantification',\n",
       "  'experimental_strategy': 'RNA-Seq',\n",
       "  'file_size': 4243912},\n",
       " {'id': 'a04267ed-bdc2-485a-8836-106a044a2bdf',\n",
       "  'data_format': 'TSV',\n",
       "  'file_name': 'TCGA-BLCA.a2cfc450-4766-43ef-9d77-3fbb1f0f431a.ascat3.gene_level_copy_number.v36.tsv',\n",
       "  'file_id': 'a04267ed-bdc2-485a-8836-106a044a2bdf',\n",
       "  'data_type': 'Gene Level Copy Number',\n",
       "  'experimental_strategy': 'Genotyping Array',\n",
       "  'file_size': 3445505},\n",
       " {'id': '031892cf-49d1-4029-91d9-f25ffc39cbac',\n",
       "  'data_format': 'SVS',\n",
       "  'file_name': 'TCGA-XF-AAN0-01A-01-TSA.1A8CAEE2-EFB6-4DA3-B6DD-F0C70AAD8CAB.svs',\n",
       "  'file_id': '031892cf-49d1-4029-91d9-f25ffc39cbac',\n",
       "  'data_type': 'Slide Image',\n",
       "  'experimental_strategy': 'Tissue Slide',\n",
       "  'file_size': 267829071},\n",
       " {'id': '98ec632d-c5a2-42cb-af09-5d2aadc54693',\n",
       "  'data_format': 'TSV',\n",
       "  'file_name': 'TCGA-BLCA.a2cfc450-4766-43ef-9d77-3fbb1f0f431a.absolute_liftover.gene_level_copy_number.v36.tsv',\n",
       "  'file_id': '98ec632d-c5a2-42cb-af09-5d2aadc54693',\n",
       "  'data_type': 'Gene Level Copy Number',\n",
       "  'experimental_strategy': 'Genotyping Array',\n",
       "  'file_size': 3436679}]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_case_metadata(SAMPLE_CASE_ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_file(file_id, file_type, dir):\n",
    "    url = f\"https://api.gdc.cancer.gov/data/{file_id}\"\n",
    "    path = os.path.join(os.getcwd(), dir, file_id + f'.{file_type.lower()}')\n",
    "    \n",
    "    # Make the request to the GDC API\n",
    "    response = requests.get(url, stream=True)\n",
    "    \n",
    "    # Check if the request was successful\n",
    "    if response.status_code == 200:\n",
    "        # Write the file content to the specified path\n",
    "        with open(path, 'wb') as file:\n",
    "            for chunk in response.iter_content(chunk_size=8192):\n",
    "                file.write(chunk)\n",
    "        print(f\"File downloaded successfully to {path}\")\n",
    "    else:\n",
    "        response.raise_for_status()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File downloaded successfully to /Users/adobles/stanford-classes/cs286/data/ac728bb8-2b0c-4a0a-9c80-23e0fb768393.tsv\n"
     ]
    }
   ],
   "source": [
    "download_file('ac728bb8-2b0c-4a0a-9c80-23e0fb768393', 'tsv', 'data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_exp = pd.read_csv('data/ac728bb8-2b0c-4a0a-9c80-23e0fb768393.tsv', sep='\\t', header=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>unstranded</th>\n",
       "      <th>stranded_first</th>\n",
       "      <th>stranded_second</th>\n",
       "      <th>tpm_unstranded</th>\n",
       "      <th>fpkm_unstranded</th>\n",
       "      <th>fpkm_uq_unstranded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N_unmapped</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1567087</td>\n",
       "      <td>1567087</td>\n",
       "      <td>1567087</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>N_multimapping</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6095774</td>\n",
       "      <td>6095774</td>\n",
       "      <td>6095774</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N_noFeature</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4002155</td>\n",
       "      <td>36415455</td>\n",
       "      <td>36630134</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N_ambiguous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6810946</td>\n",
       "      <td>1630264</td>\n",
       "      <td>1612558</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000003.15</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2297</td>\n",
       "      <td>1183</td>\n",
       "      <td>1114</td>\n",
       "      <td>22.6716</td>\n",
       "      <td>8.3908</td>\n",
       "      <td>9.2281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60659</th>\n",
       "      <td>ENSG00000288669.1</td>\n",
       "      <td>AC008763.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60660</th>\n",
       "      <td>ENSG00000288670.1</td>\n",
       "      <td>AL592295.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>630</td>\n",
       "      <td>339</td>\n",
       "      <td>333</td>\n",
       "      <td>15.6090</td>\n",
       "      <td>5.7769</td>\n",
       "      <td>6.3534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60661</th>\n",
       "      <td>ENSG00000288671.1</td>\n",
       "      <td>AC006486.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60662</th>\n",
       "      <td>ENSG00000288674.1</td>\n",
       "      <td>AL391628.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0465</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>0.0189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60663</th>\n",
       "      <td>ENSG00000288675.1</td>\n",
       "      <td>AP006621.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>34</td>\n",
       "      <td>42</td>\n",
       "      <td>41</td>\n",
       "      <td>0.9061</td>\n",
       "      <td>0.3353</td>\n",
       "      <td>0.3688</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60664 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id   gene_name       gene_type  unstranded  \\\n",
       "0              N_unmapped         NaN             NaN     1567087   \n",
       "1          N_multimapping         NaN             NaN     6095774   \n",
       "2             N_noFeature         NaN             NaN     4002155   \n",
       "3             N_ambiguous         NaN             NaN     6810946   \n",
       "4      ENSG00000000003.15      TSPAN6  protein_coding        2297   \n",
       "...                   ...         ...             ...         ...   \n",
       "60659   ENSG00000288669.1  AC008763.4  protein_coding           0   \n",
       "60660   ENSG00000288670.1  AL592295.6          lncRNA         630   \n",
       "60661   ENSG00000288671.1  AC006486.3  protein_coding           0   \n",
       "60662   ENSG00000288674.1  AL391628.1  protein_coding          10   \n",
       "60663   ENSG00000288675.1  AP006621.6  protein_coding          34   \n",
       "\n",
       "       stranded_first  stranded_second  tpm_unstranded  fpkm_unstranded  \\\n",
       "0             1567087          1567087             NaN              NaN   \n",
       "1             6095774          6095774             NaN              NaN   \n",
       "2            36415455         36630134             NaN              NaN   \n",
       "3             1630264          1612558             NaN              NaN   \n",
       "4                1183             1114         22.6716           8.3908   \n",
       "...               ...              ...             ...              ...   \n",
       "60659               0                0          0.0000           0.0000   \n",
       "60660             339              333         15.6090           5.7769   \n",
       "60661               0                0          0.0000           0.0000   \n",
       "60662               4                6          0.0465           0.0172   \n",
       "60663              42               41          0.9061           0.3353   \n",
       "\n",
       "       fpkm_uq_unstranded  \n",
       "0                     NaN  \n",
       "1                     NaN  \n",
       "2                     NaN  \n",
       "3                     NaN  \n",
       "4                  9.2281  \n",
       "...                   ...  \n",
       "60659              0.0000  \n",
       "60660              6.3534  \n",
       "60661              0.0000  \n",
       "60662              0.0189  \n",
       "60663              0.3688  \n",
       "\n",
       "[60664 rows x 9 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gene_id', 'gene_name', 'gene_type', 'unstranded', 'stranded_first',\n",
       "       'stranded_second', 'tpm_unstranded', 'fpkm_unstranded',\n",
       "       'fpkm_uq_unstranded'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_exp.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_type\n",
       "protein_coding                        0.329058\n",
       "lncRNA                                0.278600\n",
       "processed_pseudogene                  0.167595\n",
       "unprocessed_pseudogene                0.043090\n",
       "misc_RNA                              0.036463\n",
       "snRNA                                 0.031337\n",
       "miRNA                                 0.031007\n",
       "TEC                                   0.017424\n",
       "snoRNA                                0.015545\n",
       "transcribed_unprocessed_pseudogene    0.015479\n",
       "transcribed_processed_pseudogene      0.008242\n",
       "rRNA_pseudogene                       0.008193\n",
       "IG_V_pseudogene                       0.003083\n",
       "IG_V_gene                             0.002390\n",
       "transcribed_unitary_pseudogene        0.002275\n",
       "TR_V_gene                             0.001747\n",
       "unitary_pseudogene                    0.001615\n",
       "TR_J_gene                             0.001302\n",
       "scaRNA                                0.000808\n",
       "polymorphic_pseudogene                0.000791\n",
       "rRNA                                  0.000775\n",
       "IG_D_gene                             0.000610\n",
       "TR_V_pseudogene                       0.000544\n",
       "Mt_tRNA                               0.000363\n",
       "pseudogene                            0.000297\n",
       "IG_J_gene                             0.000297\n",
       "IG_C_gene                             0.000231\n",
       "IG_C_pseudogene                       0.000148\n",
       "ribozyme                              0.000132\n",
       "TR_C_gene                             0.000099\n",
       "sRNA                                  0.000082\n",
       "NaN                                   0.000066\n",
       "TR_D_gene                             0.000066\n",
       "TR_J_pseudogene                       0.000066\n",
       "IG_J_pseudogene                       0.000049\n",
       "translated_processed_pseudogene       0.000033\n",
       "Mt_rRNA                               0.000033\n",
       "scRNA                                 0.000016\n",
       "vault_RNA                             0.000016\n",
       "translated_unprocessed_pseudogene     0.000016\n",
       "IG_pseudogene                         0.000016\n",
       "Name: proportion, dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_exp.gene_type.value_counts(normalize=True, dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Level Copy Nums"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File downloaded successfully to /Users/adobles/stanford-classes/cs286/data/a04267ed-bdc2-485a-8836-106a044a2bdf.tsv\n"
     ]
    }
   ],
   "source": [
    "download_file('a04267ed-bdc2-485a-8836-106a044a2bdf', 'tsv', 'data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "copy_num = pd.read_csv('data/a04267ed-bdc2-485a-8836-106a044a2bdf.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>copy_number</th>\n",
       "      <th>min_copy_number</th>\n",
       "      <th>max_copy_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60618</th>\n",
       "      <td>ENSG00000124334.17_PAR_Y</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57184101</td>\n",
       "      <td>57197337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60619</th>\n",
       "      <td>ENSG00000270726.6_PAR_Y</td>\n",
       "      <td>AJ271736.1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57190738</td>\n",
       "      <td>57208756</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60620</th>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57203357</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60621</th>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57207346</td>\n",
       "      <td>57212230</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60622</th>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57212184</td>\n",
       "      <td>57214397</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60623 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_id    gene_name chromosome     start       end  \\\n",
       "0             ENSG00000223972.5      DDX11L1       chr1     11869     14409   \n",
       "1             ENSG00000227232.5       WASH7P       chr1     14404     29570   \n",
       "2             ENSG00000278267.1    MIR6859-1       chr1     17369     17436   \n",
       "3             ENSG00000243485.5  MIR1302-2HG       chr1     29554     31109   \n",
       "4             ENSG00000284332.1    MIR1302-2       chr1     30366     30503   \n",
       "...                         ...          ...        ...       ...       ...   \n",
       "60618  ENSG00000124334.17_PAR_Y         IL9R       chrY  57184101  57197337   \n",
       "60619   ENSG00000270726.6_PAR_Y   AJ271736.1       chrY  57190738  57208756   \n",
       "60620  ENSG00000185203.12_PAR_Y       WASIR1       chrY  57201143  57203357   \n",
       "60621  ENSG00000182484.15_PAR_Y       WASH6P       chrY  57207346  57212230   \n",
       "60622   ENSG00000227159.8_PAR_Y     DDX11L16       chrY  57212184  57214397   \n",
       "\n",
       "       copy_number  min_copy_number  max_copy_number  \n",
       "0              NaN              NaN              NaN  \n",
       "1              NaN              NaN              NaN  \n",
       "2              NaN              NaN              NaN  \n",
       "3              NaN              NaN              NaN  \n",
       "4              NaN              NaN              NaN  \n",
       "...            ...              ...              ...  \n",
       "60618          NaN              NaN              NaN  \n",
       "60619          NaN              NaN              NaN  \n",
       "60620          NaN              NaN              NaN  \n",
       "60621          NaN              NaN              NaN  \n",
       "60622          NaN              NaN              NaN  \n",
       "\n",
       "[60623 rows x 8 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "copy_number\n",
       "2.0    0.583953\n",
       "3.0    0.202481\n",
       "1.0    0.192122\n",
       "NaN    0.013246\n",
       "4.0    0.008149\n",
       "6.0    0.000049\n",
       "Name: proportion, dtype: float64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_num.copy_number.value_counts(normalize=True, dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "max_copy_number\n",
       "2.0    0.583772\n",
       "3.0    0.202679\n",
       "1.0    0.192056\n",
       "NaN    0.013246\n",
       "4.0    0.008182\n",
       "6.0    0.000066\n",
       "Name: proportion, dtype: float64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_num.max_copy_number.value_counts(normalize=True, dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "min_copy_number\n",
       "2.0    0.584135\n",
       "3.0    0.202332\n",
       "1.0    0.192171\n",
       "NaN    0.013246\n",
       "4.0    0.008066\n",
       "6.0    0.000049\n",
       "Name: proportion, dtype: float64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_num.min_copy_number.value_counts(normalize=True, dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gene_id', 'gene_name', 'chromosome', 'start', 'end', 'copy_number',\n",
       "       'min_copy_number', 'max_copy_number'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_num.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### WSIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File downloaded successfully to /Users/adobles/stanford-classes/cs286/data/67e5f3a1-c5f8-46b5-a084-71378cd3dfd6.svs\n"
     ]
    }
   ],
   "source": [
    "download_file('67e5f3a1-c5f8-46b5-a084-71378cd3dfd6', 'svs', 'data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': '32e5edd4-51c8-4c6f-b58b-2387d97ee546',\n",
       "  'data_format': 'TSV',\n",
       "  'file_name': 'TCGA-LUAD.039dd88e-a3dc-4abd-a8b1-9cfcfe98842f.gene_level_copy_number.v36.tsv',\n",
       "  'file_id': '32e5edd4-51c8-4c6f-b58b-2387d97ee546',\n",
       "  'data_type': 'Gene Level Copy Number',\n",
       "  'file_size': 3446825},\n",
       " {'id': '5f739d40-1c4f-47f4-87d1-4368e8b6b0b9',\n",
       "  'data_format': 'SVS',\n",
       "  'file_name': 'TCGA-55-A48X-01Z-00-DX1.A46C6373-8458-4D55-88C3-4C70A05F9F47.svs',\n",
       "  'file_id': '5f739d40-1c4f-47f4-87d1-4368e8b6b0b9',\n",
       "  'data_type': 'Slide Image',\n",
       "  'file_size': 50495298},\n",
       " {'id': 'b06a3383-3a16-4f91-91c4-71cad60dae4b',\n",
       "  'data_format': 'TSV',\n",
       "  'file_name': '897e33bc-3862-47bc-8277-a9b1b5371802.rna_seq.augmented_star_gene_counts.tsv',\n",
       "  'file_id': 'b06a3383-3a16-4f91-91c4-71cad60dae4b',\n",
       "  'data_type': 'Gene Expression Quantification',\n",
       "  'file_size': 4245425},\n",
       " {'id': 'e4e0fe93-936a-49d6-8119-b08ccf644169',\n",
       "  'data_format': 'TSV',\n",
       "  'file_name': 'TCGA-LUAD.039dd88e-a3dc-4abd-a8b1-9cfcfe98842f.ascat3.gene_level_copy_number.v36.tsv',\n",
       "  'file_id': 'e4e0fe93-936a-49d6-8119-b08ccf644169',\n",
       "  'data_type': 'Gene Level Copy Number',\n",
       "  'file_size': 3445505},\n",
       " {'id': '68d224ad-2217-48ab-a914-d482affa6957',\n",
       "  'data_format': 'SVS',\n",
       "  'file_name': 'TCGA-55-A48X-01A-01-TS1.8A820B15-6DF2-48FC-A539-0BA306166C9C.svs',\n",
       "  'file_id': '68d224ad-2217-48ab-a914-d482affa6957',\n",
       "  'data_type': 'Slide Image',\n",
       "  'file_size': 112920366},\n",
       " {'id': '2966dc66-47d5-41d2-bebf-e92e33a410bb',\n",
       "  'data_format': 'TSV',\n",
       "  'file_name': 'TCGA-LUAD.039dd88e-a3dc-4abd-a8b1-9cfcfe98842f.absolute_liftover.gene_level_copy_number.v36.tsv',\n",
       "  'file_id': '2966dc66-47d5-41d2-bebf-e92e33a410bb',\n",
       "  'data_type': 'Gene Level Copy Number',\n",
       "  'file_size': 3434420}]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_case_metadata(cohort.loc[0, 'id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cancer",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
